### PR TITLE
feat: add getLockFileGuard method to util

### PR DESCRIPTION
**Pull request by jlipps**
the idea is to support guarding behaviors across multiple processes: https://github.com/appium/appium/pull/13463#pullrequestreview-304978624

### CODE REVIEW

**Comment by mykola-mokhnach** [comment]
I would rather use some ready solution like https://www.npmjs.com/package/lockfile and create a wrapper over it inside appium-support

File: lib/fs.js [hunk 1]
```diff
@@ -158,6 +159,8 @@ for (const s of simples) {
 const syncFunctions = [
   'createReadStream',
   'createWriteStream',
+  'writeFileSync',
```
Comment (mykola-mokhnach) [inline lib/fs.js:162]: so we still need these?

File: lib/util.js [hunk 2]
```diff
@@ -443,11 +444,49 @@ async function toInMemoryBase64 (srcPath, opts = {}) {
...
+ * Create an async function which, when called, will not proceed until a certain file is no
+ * longer present on the system. This allows for preventing concurrent behavior across processes
+ * using a known lockfile path.
+ *
+ * @param {string} lockFile The full path to the file used for the lock
+ * @param {number} timeout [120] The max time in seconds to wait for the lock
+ * @returns {AsyncFunction} async function that takes another async function defining the locked
+ * behavior
+ */
+function getLockFileGuard (lockFilePath, timeout = 120) {
```
Comment (mykola-mokhnach) [inline lib/util.js:457]: it makes sense to make options a dictionary, so we can add more of them later

File: lib/util.js [hunk 3]
```diff
@@ -443,11 +444,49 @@ async function toInMemoryBase64 (srcPath, opts = {}) {
...
+  const lock = B.promisify(lockfile.lock);
+  const checkSync = B.promisify(lockfile.checkSync);
+  const unlock = B.promisify(lockfile.unlock);
+
+  return async (behavior) => {
+    try {
+      // if the lockfile doesn't exist, lock it synchronously to make sure no other call
+      // on the same spin of the event loop can also initiate a lock. If the lockfile does exist
+      // then just use the regular async 'lock' method which will wait on the lock.
+      if (!checkSync(lockFilePath)) {
```
Comment (mykola-mokhnach) [inline lib/util.js:468]: this call is async

File: lib/util.js [hunk 4]
```diff
@@ -443,11 +444,49 @@ async function toInMemoryBase64 (srcPath, opts = {}) {
...
+ * longer present on the system. This allows for preventing concurrent behavior across processes
+ * using a known lockfile path.
+ *
+ * @param {string} lockFile The full path to the file used for the lock
+ * @param {number} timeout [120] The max time in seconds to wait for the lock
+ * @returns {AsyncFunction} async function that takes another async function defining the locked
+ * behavior
+ */
+function getLockFileGuard (lockFilePath, timeout = 120) {
+  const lockSync = B.promisify(lockfile.lockSync);
```
Comment (mykola-mokhnach) [inline lib/util.js:458]: something is not right here. How lockSync can be promisified?

**Comment by jlipps** [comment]
comments addressed!

File: lib/util.js [hunk 5]
```diff
@@ -443,11 +444,51 @@ async function toInMemoryBase64 (srcPath, opts = {}) {
...
+/**
+ * Create an async function which, when called, will not proceed until a certain file is no
+ * longer present on the system. This allows for preventing concurrent behavior across processes
+ * using a known lockfile path.
+ *
+ * @param {LockFileOptions} opts
+ * @returns {AsyncFunction} async function that takes another async function defining the locked
+ * behavior
+ */
+function getLockFileGuard ({lockFile, timeout = 120}) {
```
Comment (mykola-mokhnach) [inline lib/util.js:461]: the first parameter is always required. I meant something like

```
async function getLockFileGuard (lockFile, opts = {})
```

Comment (jlipps) [reply lib/util.js:461]: oh i see, right.

File: test/util-e2e-specs.js [hunk 6]
```diff
@@ -33,4 +34,103 @@ describe('#util', function () {
     });
   });
 
+  describe('getLockFileGuard()', function () {
+    let tmpRoot;
+    let lockFile;
+    let testFile;
+
+    async function lockExists () {
```
Comment (mykola-mokhnach) [inline test/util-e2e-specs.js:42]: should we use appropriate methods from `lockfile` module instead or have wrappers over them to check if lock exists?

Comment (jlipps) [reply test/util-e2e-specs.js]: yeah, good idea.

File: lib/fs.js [hunk 7]
```diff
@@ -150,6 +150,7 @@ let fs = {
 const simples = [
   'open', 'close', 'access', 'readFile', 'writeFile', 'write', 'read',
   'readlink', 'chmod', 'unlink', 'readdir', 'stat', 'rename', 'lstat',
+  'appendFile'
```
Comment (mykola-mokhnach) [inline lib/fs.js:153]: do we need this?

File: test/util-e2e-specs.js [hunk 8]
```diff
@@ -33,4 +35,95 @@ describe('#util', function () {
     });
   });
 
+  describe('getLockFileGuard()', function () {
+    let tmpRoot;
+    let lockFile;
+    let testFile;
+    const lockExists = () => B.promisify(lockfile.check)(lockFile);
```
Comment (mykola-mokhnach) [inline test/util-e2e-specs.js:42]: does it make sense to make this check public so we could check the lock in other modules without importing lockfile explicitly?

Comment (jlipps) [reply test/util-e2e-specs.js]: yeah, good idea. done.

**Pull request merged**